lamindb/dev/db/_add.py

```diff
@@ -176,32 +176,35 @@ def local_instance_storage_matches_local_parent(dobject: DObject):
     return str(storage.root) in parents
 
 
-def write_objectkey(record) -> None:
+def get_storage_root_str():
+    root = setup_settings.instance.storage.root
+    root_str = str(root)
+    if isinstance(root, UPath):
+        root_str = root_str.rstrip("/")
+    return root_str
+
+
+def write_objectkey(record: sqm.SQLModel) -> None:
     """Write to _objectkey.
 
     An objectkey excludes the storage root and the file suffix.
     """
-    storage = setup_settings.instance.storage
 
-    def set_objectkey(record: sqm.SQLModel, filepath: Union[Path, UPath]):
-        if not isinstance(filepath, UPath):  # is local filepath
-            filepath_str = filepath.resolve().as_posix()
-        else:  # is remote path
-            filepath_str = filepath.as_posix()
-        root_str = storage.root.as_posix()
-        if root_str[-1] != "/":
-            root_str += "/"
-
-        if not isinstance(record, (DObject, DFolder)):
-            raise NotImplementedError
-        # for DObject, _objectkey is relative path without suffix
+    def set_objectkey(record: Union[DObject, DFolder], filepath: Union[Path, UPath]):
+        root_str = get_storage_root_str()
+
+        try:  # is local filepath
+            relpath = filepath.relative_to(root_str)
```

Comment:
I think it is better to test for `UPath` class instead of relying on `try except` to figure out if it is cloud storage.

Reply:
Fixed, and fixed another occurance of str(path)